### ISSUE TITLE
Problem with a pls station

### ISSUE THREAD

**Issue by MaDill** [y20k/transistor#82]
This station is not working: 
http://mp3stream1.apasf.apa.at:8000/listen.pls

Here are some i fos about the radio station http://fm4.orf.at/stream

Tested with play Store Version 2.0

**Comment by y20k** [comment]
The station you mentioned  works on my device (adding and playback). The station's name initially is just "listen". It "should" work on your phone, too.

I am on vacation right now. I can look into it next week. 

In the the meantime you might want to have a look into the folder /Android/data/org.y20k.transistor/Collection using a file manager and see if there are duplicate files etc. Maybe there is a station named listen.m3u.

**Comment by y20k** [comment]
PS. You could try to add the raw audio stream using the +symbol ==> http://mp3stream1.apasf.apa.at:8000/

**Comment by MaDill** [comment]
Thank you for the feedback and enjoy your vacation. I tried to add again the channel with the URL, but then transistor crash immediately. I see in the google error report, that I use version 2.0.1
When I create the file manually in the folder, it show up in the list. When I want to play it, the "LED" turns to green but no sound. I installed a tool that visualize the transfer rate and it show 0. Other channels works fine.

**Comment by y20k** [comment]
I am able to recreate the crash. That should not happen. It occurs when I try to add that specific URL for the second time. I will try to catch that crash in the next version. It is not related to your main problem though.

That problem remains: You can add the station (somehow :-) ), but it does not play back. That is very strange. The stream is just a plain MP3@96kb/s. I plays on my device.

On which device are you trying to listen. Which Android Version? A custom ROM? Does the device have a SD card.

**Comment by MaDill** [comment]
I use a Google Nexus 7 with 3G from 2012. Android Version is 4.4.4. No modifications, except root. I can play the downloaded pls playlist with the VLC player. Then I can listen to my favorite station, but not in transistor.

When I try to open the file from total commander I get following error. It's in German and means that transistor has no idea how to handle the file.
![screenshot_2016-04-29-22-26-01](https://cloud.githubusercontent.com/assets/7298934/14928712/e9ef8f42-0e59-11e6-983e-919f7ff3f88e.png)

**Comment by y20k** [comment]
Hi there,

as far as I can see there two problems:

**1) Station does not play back**
That is strange, because I can listen to FM4 on my device. Further debugging: You can tap on the station name in the Now Playing screen. That should copy the station URL that Transistor has stored to the clipboard. Compare it to http://mp3stream1.apasf.apa.at:8000/ - is it identical?

**2) Unable to add station**
Judging from your screenshot you try to add a station that is already in the correct `Collection` folder. That should be unnecessary. Transistor should be able to read the station if it has the extension m3u and is formatted correctly (see below). 
If on the other hand you want to add a new station using a file manager, put this file outside of the `Collection` folder - and make sure that there is not already a m3u-file for that station in `Collection`. 
Transistor will (a) read this file (pls or m3u), (b) then check if it contains a valid streaming URL (requires network connection), (c) then write a m3u file into the `Collection` folder.
In short: Before you try something, make sure that there is no `FM4.m3u` (or `listen.m3u`) present in the `Collection` folder.

==> **FM4.m3u**
```
#EXTM3U

#EXTINF:-1,FM4
http://mp3stream1.apasf.apa.at:8000/

```

**Comment by MaDill** [comment]
Hi,
thank you for trying to help me. I tested your points.
to 1) I tap on the station and paste it in my notes. It show: listen(http://mp3stream1.apasf.apa.at:8000/) Except the "listen(" at the beginning and the ")" at the end it match.
to 2) I deleted the file from the Collection folder and clicked it on the filemanager to play it with transistor. I receive the same error, just as file content it show the content of the downloaded "listen.pls" from here http://mp3stream1.apasf.apa.at:8000/listen.pls

3) I tried to copy the pls file in the collection folder. It doesn't show up, when I start transistor. When I rename it to m3u, it is shown in the list in transistor, but with the same result (LED turns green, but no sound).

**Comment by y20k** [comment]
> LED turns green, but no sound

This is the part that I find very irritating. Unfortunately I do not have an easy answer. Transistor uses Android's [MediaPlayer](http://developer.android.com/reference/android/media/MediaPlayer.html). VLC on the other hand has it's own audio engine. Some streams play in VLC and not in Transistor.

But in this case: The [FM4 steam](http://mp3stream1.apasf.apa.at:8000/) plays on my Nexus 5 and it does not play on your Nexus 7. Both run on stock Android Versions straight from Google. The stream itself is a very simple and common MP3 stream. 

Maybe it is the different Android Version. I will to test that using an emulator in the next couple of days.

**Comment by y20k** [comment]
Hi  @MaDill 

today I got to analyze this issue using an emulator. I was able to reproduce the whole thing on an emulated Galaxy Nexus running Android 4.4. Here is what I found:

**A) Unable to add station**
There is no way to add the FM4 station (http://mp3stream1.apasf.apa.at:8000/listen.pls) you mentioned the usual way on Android 4.4. Transistor checks the PLS itself as well the URL with the PLS for the type of content. 

On **Android 4.4** the log looks like this
```
/org.y20k.transistor V/Station: Content type of given file is ContentType{type='audio/x-scpls', charset='null'}
/org.y20k.transistor V/Station: Downloading... http://mp3stream1.apasf.apa.at:8000/listen.pls
/org.y20k.transistor V/Station: Content type of URL within playlist: null
```

On **Android 6** the log looks a little bit different
```
/org.y20k.transistor V/Station: Content type of given file is ContentType{type='audio/x-scpls', charset='null'}
/org.y20k.transistor V/Station: Downloading... http://mp3stream1.apasf.apa.at:8000/listen.pls
/org.y20k.transistor V/Station: Content type of URL within playlist: ContentType{type='audio/mpeg', charset='null'}
```

Transistor will refuse to store a station without a content type or with the wrong content type. 
`Content type of URL within playlist: null` ==> Transistor rejects this station on Android 4.4. 
Transistor has a [getContentType](https://github.com/y20k/transistor/blob/master/app/src/main/java/org/y20k/transistor/core/Station.java#L264) method that returns `null` if certain conditions are not met. Somehow it behaves differently on Android 4.4 (@meonwax : do you have any insight on this issue?).

**B) Station does not play back**
I can bypass the above mentioned content type check if I manually place an M3U file containing the "real" stream URL (http://mp3stream1.apasf.apa.at:8000/) into Transistor's Collection folder. But then on Android 4.4 Transistor's UI falsely shows a green light (as you mentioned).  

A look at the log shows a playback error caused by the Android [MediaPlayer](https://developer.android.com/reference/android/media/MediaPlayer.html):
```
/org.y20k.transistor E/MediaPlayer: error (1, -1004)
/org.y20k.transistor E/MediaPlayer: Error (1,-1004)
/org.y20k.transistor E/PlayerService: Unknown media playback error
/org.y20k.transistor E/PlayerService: IO media error.

```

**My conclusion (so far):**
It maybe possible to fix **Unable to add station** on Transistor's side. But I do not see the benefit if in the end [MediaPlayer](https://developer.android.com/reference/android/media/MediaPlayer.html) still rejects the streaming URL. I will continue to look around on the web for similar problems - but I do not have too much hope for now.

**Comment by MaDill** [comment]
OK. Anyway. Thank you for the support and your help to find the problem. Before I used https://github.com/yomguy/servestream to play my radio streams. But it looks like at this project is no active development. This app is able to play the stream. I don't know how this app handle to play streams, but maybe you can have a look.

**Comment by y20k** [comment]
Servestream seems to have [multiple audio players](https://github.com/yomguy/servestream/tree/master/src/net/sourceforge/servestream/media). 

* [NativePlayer.java](https://github.com/yomguy/servestream/blob/master/src/net/sourceforge/servestream/media/NativePlayer.java) uses the same Android MediaPlayer that Transistor uses. 
* [FFmpegMediaPlayer.java](https://github.com/yomguy/servestream/blob/master/src/net/sourceforge/servestream/media/FFmpegMediaPlayer.java) uses ... FFmpeg.

I did not dig in too deep into it to see which one is really used, of if there is a method that decides which one to use.

I try to keep Transistor very simple. Not only on the UI side, but also code-wise. That is because I am afraid of maintaining a large code base. I do not have too much time I can spend on it and I am only a semi-experienced coder.

So, if not absolutely necessary I would never build two different audio players. I even hesitate to switch from MediaPlayer to another engine like FFmpeg. 

Therefore I guess I will close this issue - unresolved as it is. Sorry @MaDill - and thank you for sticking with it and helping me to understand the issue. I learned something.
